### ISSUE TITLE
Load PeftModel in another PretrainedModel

### ISSUE THREAD

**Issue by 648652443** [huggingface/peft#1042]
### System Info

torch                        2.0.1
transformers                 4.34.0
peft                         0.5.0

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

When I load PeftModel in my PretrainedModel, it will cause some question.
If I load PeftModel by this:
```python
        peft_model_id = "llama2_all0.1_llama2-7B-hf_clip_lr0.0001_bs240_ip1536_op1664_ep3/checkpoint-39006"

        peft_config = PeftConfig.from_pretrained(peft_model_id)

        language_model = LlamaForCausalLM.from_pretrained(
            peft_config.base_model_name_or_path,
            load_in_8bit=True,
            device_map="auto",
            torch_dtype=torch.float16
        )

        language_model = PeftModel(language_model, peft_config)
```
The traceback is:

```
> Traceback (most recent call last):
  File "main_central_bllama.py", line 337, in <module>
    model = Blip2ForConditionalGeneration.from_pretrained("blip2llamaConfig", device_map="auto", torch_dtype=torch.float16)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/transformers/modeling_utils.py", line 3085, in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
  File "/home/sqzhang/blip2Agent/modeling_blip_2.py", line 1608, in __init__
    language_model = PeftModel(language_model, peft_config)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/peft_model.py", line 111, in __init__
    self.base_model = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type](
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/lora.py", line 274, in __init__
    super().__init__(model, config, adapter_name)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/tuners_utils.py", line 88, in __init__
    self.inject_adapter(self.model, adapter_name)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/tuners_utils.py", line 219, in inject_adapter
    self._create_and_replace(peft_config, adapter_name, target, target_name, parent, **optionnal_kwargs)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/lora.py", line 373, in _create_and_replace
    self._replace_module(parent, target_name, new_module, target)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/lora.py", line 378, in _replace_module
    new_module.weight = child.weight
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1632, in __setattr__
    self.register_parameter(name, value)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/accelerate/big_modeling.py", line 123, in register_empty_parameter
    module._parameters[name] = param_cls(module._parameters[name].to(device), **kwargs)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/bitsandbytes/nn/modules.py", line 288, in __new__
    return torch.Tensor._make_subclass(cls, data, requires_grad)
RuntimeError: Only Tensors of floating point and complex dtype can require gradients
```

Or if I load PeftModel by this:
```python
        peft_model_id = "llama2_all0.1_llama2-7B-hf_clip_lr0.0001_bs240_ip1536_op1664_ep3/checkpoint-39006"

        peft_config = PeftConfig.from_pretrained(peft_model_id)

        language_model = LlamaForCausalLM.from_pretrained(
            peft_config.base_model_name_or_path,
            device_map="auto",
            torch_dtype=torch.float16
        )

        language_model = PeftModel(language_model, peft_config)
```
The Traceback is:

```
> Traceback (most recent call last):
  File "main_central_bllama.py", line 337, in <module>
    model = Blip2ForConditionalGeneration.from_pretrained("blip2llamaConfig", device_map="auto", torch_dtype=torch.float16)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/transformers/modeling_utils.py", line 3085, in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
  File "/home/sqzhang/blip2Agent/modeling_blip_2.py", line 1608, in __init__
    language_model = PeftModel(language_model, peft_config)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/peft_model.py", line 111, in __init__
    self.base_model = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type](
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/lora.py", line 274, in __init__
    super().__init__(model, config, adapter_name)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/tuners_utils.py", line 88, in __init__
    self.inject_adapter(self.model, adapter_name)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/tuners_utils.py", line 219, in inject_adapter
    self._create_and_replace(peft_config, adapter_name, target, target_name, parent, **optionnal_kwargs)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/lora.py", line 373, in _create_and_replace
    self._replace_module(parent, target_name, new_module, target)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/peft/tuners/lora.py", line 390, in _replace_module
    module.to(child.weight.device)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1145, in to
    return self._apply(convert)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/torch/nn/modules/module.py", line 797, in _apply
    module._apply(fn)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/torch/nn/modules/module.py", line 820, in _apply
    param_applied = fn(param)
  File "/home/sqzhang/anaconda3/envs/agent/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1143, in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
NotImplementedError: Cannot copy out of meta tensor; no data!
```

### Expected behavior

How can I load PeftModel in my PretrainedModel

**Comment by 648652443** [comment]
> Instead of calling `PeftModel(...)` directly, please always use `get_peft_model`. I'm not sure if this will solve your problem, but could you please give it a try?

Thanks for your advice! I had already tried this method and faced same question.
But now I have loaded the peft model by this:
`
_no_split_modules = ["LlamaDecoderLayer", "LlamaForCausalLM"]

def __init__(self, config: Config):

        peft_model_id = "llama2_all0.1_llama2-7B-hf_clip_lr0.0001_bs240_ip1536_op1664_ep3/checkpoint-39006"
        peft_config = PeftConfig.from_pretrained(peft_model_id)
        # Update _tied_weights_keys using the base model used.
        language_model = AutoModelForCausalLM.from_config(config.text_config)
        if language_model._tied_weights_keys is not None:
            self._tied_weights_keys = [f"language_model.{k}" for k in language_model._tied_weights_keys]
        language_model = PeftModel(language_model, peft_config)

        self.language_model = language_model
`

**Comment by BenjaminBossan** [comment]
So is it working now with your fix?

**Comment by 648652443** [comment]
> So is it working now with your fix?

Yes. But I'm facing another question beacuse of the function init_post(). My peft model weights were initialized. So I'm looking for method to solve it. hahaha

**Comment by LuciferianInk** [comment]
I am currently running into the same error:
```
RuntimeError: only Tensors of floating point and complex dtype can require gradients
```
This happens when I try to set "requires_grad" on PEFT layers:
```
model = get_peft_model(model, peft_config)
model = prepare_model_for_kbit_training(model)
for name, param in model.named_parameters():
    if any(l in name.lower() for l in ["lora", "lokr", "ia3", "base_layer"]):
        param.requires_grad = True
```
If I remove these calls to "requires_grad", then the output of `model.print_trainable_parameters()` is 0.
```
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]

  | Name  | Type                 | Params
-----------------------------------------------
0 | model | PeftModelForCausalLM | 75.7 M
-----------------------------------------------
0         Trainable params
75.7 M    Non-trainable params
75.7 M    Total params
302.890   Total estimated model params size (MB)
```

Of course, with 0 trainable parameters, my model fails to learn.

So, I'm not really sure what to do here. Quantized training does not work because of this error - and training (in general) doesn't work when "requires_grad" is False for PEFT layers.

**Comment by BenjaminBossan** [comment]
@LuciferianInk What you're getting looks a bit strange. Could you provide more code to reproduce the issue?

**Comment by LuciferianInk** [comment]
Thanks, but I think I figured out the problem. I was able to upgrade to 0.6.2, and after doing so - I was able to completely remove that `requires_grad` stuff. It appears that `prepare_model_for_kbit_training()` does all of that for me, and I've been able to move forward in 4, 8 and 32 bit modes.

**Comment by younesbelkada** [comment]
Nice, I believe the issue is solved! Feel free to re-open or open a new issue if you think it is not the case. BTW we also released PEFT 0.7.1
